### ISSUE TITLE
1 psnr drop in reproducing

### ISSUE THREAD

**Issue by kaiq663** [p1y2z3/SDAP#2]
Firstly, I try my best to follow environment you provide

Secondly,  only train in SIDD_Medium without crop before (self.repeat=80, 80*320=25600)

Thirdly, lr1e-4 ps160 bs16 epochs15 on the fisrt stage. while lr1e-5 ps250 bs8 epochs10 on the second

Fouthly, finish experiments in your project

Whatever I try, there is a 1 psnr drop, SDAP 35.5055 SDAP(E) 36.5131, is there something wrong with me?Thanks

**Comment by p1y2z3** [comment]
Usually, in the first stage, SDAP achieves a PSNR result of 36.1-36.3, and training for the second stage requires loading the best model from the first stage to continue training. You can check if you load the correct model on the second stage.

**Comment by Kaiq663** [comment]
> Usually, in the first stage, SDAP achieves a PSNR result of 36.1-36.3, and training for the second stage requires loading the best model from the first stage to continue training. You can check if you load the correct model on the second stage.

the default setting of epochs at first stage in your code is 50 instead of 15 presented in paper, so I try 50 epochs in the first stage, it achieves a PSNR result of 36.5308(the best pth). The second stage is on training.

Something wrong with me or you?

**Comment by kaiq663** [comment]
> Usually, in the first stage, SDAP achieves a PSNR result of 36.1-36.3, and training for the second stage requires loading the best model from the first stage to continue training. You can check if you load the correct model on the second stage.

the default setting of epochs at first stage in your code is 50 instead of 15 presented in paper, 

Firstly, so I try 50 epochs in the first stage, and evaluate on the last checkpoint(checkpoint_epoch_0050_1599_35.66_0.8900.pth) it achieves a PSNR of 36.0458, strange! The second stage is on training.

I wonder epochs at first stage presented in your paper may be wrong?

**Comment by p1y2z3** [comment]
Have you downloaded the complete SIDD validation dataset? Validation and testing need to be conducted on the full SIDD validation dataset.

**Comment by kaiq663** [comment]
> Have you downloaded the complete SIDD validation dataset? Validation and testing need to be conducted on the full SIDD validation dataset.

During validation/test in your training, there is only five val imgs you provided.

But I choose the last pth file  to evaluate on full SIDD validation dataset, I also evaluate pretrain model named SDAP.pth you provided on full SIDD validation datset, the result is right, maybe a little permutation.

**Comment by p1y2z3** [comment]
Sorry, I didn't upload the full SIDD validation set due to github's file size limit. The selection of the optimal model needs to be done in the full SIDD validation set.

**Comment by kaiq663** [comment]
> Sorry, I didn't upload the full SIDD validation set due to github's file size limit. The selection of the optimal model needs to be done in the full SIDD validation set.

No need, I try again. By the way, nice work.

**Comment by kaiq663** [comment]
> Sorry, I didn't upload the full SIDD validation set due to github's file size limit. The selection of the optimal model needs to be done in the full SIDD validation set.

But I corporate your code into mine before which evaluate on full SIDD validation datset during training, the last pth works the best, it  still cannot achieve 36 PSNR within 15 epochs, which makes me really confused.

it achieves 2023-09-22, 07:50:06  SIDDSrgbValidationDataset, psnr: 35.1503, ssim: 0.8379 in the last pth

**Comment by p1y2z3** [comment]
There might have been an oversight on my part. In my initial training, I pre-cropped the images and set the batch size to 32. The training epochs mentioned in the paper are estimates, and the actual number may exceed the settings provided in the paper.

**Comment by yourshadow258** [comment]
Hello author, I may not be able to confirm whether the SIDD validation set will be transferred from mat to png and whether the SIDD training data cropping method is the same as yours. Could you please upload the code for dataset processing? Thank you very much. Make sure I haven't made any mistakes yet

**Comment by p1y2z3** [comment]
For issues related to processing the SIDD dataset, please refer to https://github.com/JimmyChame/SADNet. I apologize, but since I have already graduated, I cannot provide you with the specific code.

**Comment by p1y2z3** [comment]
> Hello author, I may not be able to confirm whether the SIDD validation set will be transferred from mat to png and whether the SIDD training data cropping method is the same as yours. Could you please upload the code for dataset processing? Thank you very much. Make sure I haven't made any mistakes yet

For issues related to processing the SIDD dataset, please refer to https://github.com/JimmyChame/SADNet. I apologize, but since I have already graduated, I cannot provide you with the specific code.

**Comment by yourshadow258** [comment]
Thank you for your prompt reply. This is very helpful to me. I noticed that you tested the performance of neighbor2neighbor on SIDD in a comparative experiment, which was not done in other papers. did you use the dataset you cropped? did you modify the training code of the original paper? Because my PSNR for neighbor 2 in the SIDD validation set was only 25... dropped from 26.7 to around 25, and I never had a clue why this phenomenon occurred.You have get a great performance.

**Comment by p1y2z3** [comment]
> Thank you for your prompt reply. This is very helpful to me. I noticed that you tested the performance of neighbor2neighbor on SIDD in a comparative experiment, which was not done in other papers. did you use the dataset you cropped? did you modify the training code of the original paper? Because my PSNR for neighbor 2 in the SIDD validation set was only 25... dropped from 26.7 to around 25, and I never had a clue why this phenomenon occurred.You have get a great performance.

I used their code but replaced the dataset. In fact, cropping the dataset has limited impact on performance. The poor performance of Neighbor2Neighbor is because its Gaussian noise assumption cannot be directly applied to real noise. We also observed that when Neighbor2Neighbor was trained on real noise, the PSNR consistently decreased. Therefore, we used a very short training time to achieve a relatively high PSNR.

**Comment by yourshadow258** [comment]
Thank you for the author's reply. I have indeed been troubled for a long time.

**Comment by xzq-2000** [comment]
请问一下作者，莫看您使用的BSN网络和APBSN网络使用的同一个网络，但是在训练速度上似乎比APBSN满了很多，请问一下这是我的设置的问题还是由什么原因导致的

**Comment by p1y2z3** [comment]
> 请问一下作者，莫看您使用的BSN网络和APBSN网络使用的同一个网络，但是在训练速度上似乎比APBSN满了很多，请问一下这是我的设置的问题还是由什么原因导致的

为了体现文章里的循环，我在训练代码里面对25个子图通过for循环denoising的，你想加速就写成并行呗

**Comment by xzq-2000** [comment]
我好像找到您说得的那个部分
        for j in range(p * p):
            sub_denoised[j,...] = model(noisy_sub[j,...])
我还有一个问题想要请教您，原来的APBSN中有一块代码是这样的：
def pixel_shuffle_down_sampling(x:torch.Tensor, f:int, pad:int=0, pad_value:float=0.):
#PD操作
    # single image tensor
    if len(x.shape) == 3:
        c,w,h = x.shape
        unshuffled = F.pixel_unshuffle(x, f)
        if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value)
        return unshuffled.view(c,f,f,w//f+2*pad,h//f+2*pad).permute(0,1,3,2,4).reshape(c, w+2*f*pad, h+2*f*pad)
    # batched image tensor
    else:
        b,c,w,h = x.shape
        unshuffled = F.pixel_unshuffle(x, f)
        if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value)
        return unshuffled.view(b,c,f,f,w//f+2*pad,h//f+2*pad).permute(0,1,2,4,3,5).reshape(b,c,w+2*f*pad, h+2*f*pad)
# do PD
        if pd > 1:
            pd_img = pixel_shuffle_down_sampling(img, f=pd, pad=self.pd_pad)
以上是将图片输入到APBSN之前的PD操作代码，对于(8,3,120,120)的输入在pd=5后输出依然是(8,3,120,120)，而通过您的RSG操作后形状会变成25个(8,3,24,24),我想知道的是即便相对于原版的APBSN参数量是一样的，但是运算次数变为25（每次运算的参数变为1/25），这样也会使训练时间变长对吗(感觉训练时间多了10-20倍)，辛苦您的答复

**Comment by p1y2z3** [comment]
> 我好像找到您说得的那个部分 for j in range(p * p): sub_denoised[j,...] = model(noisy_sub[j,...]) 我还有一个问题想要请教您，原来的APBSN中有一块代码是这样的： def pixel_shuffle_down_sampling(x:torch.Tensor, f:int, pad:int=0, pad_value:float=0.): #PD操作 # single image tensor if len(x.shape) == 3: c,w,h = x.shape unshuffled = F.pixel_unshuffle(x, f) if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value) return unshuffled.view(c,f,f,w//f+2_pad,h//f+2_pad).permute(0,1,3,2,4).reshape(c, w+2_f_pad, h+2_f_pad) # batched image tensor else: b,c,w,h = x.shape unshuffled = F.pixel_unshuffle(x, f) if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value) return unshuffled.view(b,c,f,f,w//f+2_pad,h//f+2_pad).permute(0,1,2,4,3,5).reshape(b,c,w+2_f_pad, h+2_f_pad)
> 
> # do PD
> ```
>     if pd > 1:
>         pd_img = pixel_shuffle_down_sampling(img, f=pd, pad=self.pd_pad)
> ```
> 
> 以上是将图片输入到APBSN之前的PD操作代码，对于(8,3,120,120)的输入在pd=5后输出依然是(8,3,120,120)，而通过您的RSG操作后形状会变成25个(8,3,24,24),我想知道的是即便相对于原版的APBSN参数量是一样的，但是运算次数变为25（每次运算的参数变为1/25），这样也会使训练时间变长对吗(感觉训练时间多了10-20倍)，辛苦您的答复

是的，就是这里导致的训练时间变长（并行要比串行快很多）

**Comment by xzq-2000** [comment]
好的好的，万分感谢



---Original---
From: ***@***.***&gt;
Date: Fri, Sep 13, 2024 11:36 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [p1y2z3/SDAP] 1 psnr drop in reproducing (Issue #2)




  
我好像找到您说得的那个部分 for j in range(p * p): sub_denoised[j,...] = model(noisy_sub[j,...]) 我还有一个问题想要请教您，原来的APBSN中有一块代码是这样的： def pixel_shuffle_down_sampling(x:torch.Tensor, f:int, pad:int=0, pad_value:float=0.): #PD操作 # single image tensor if len(x.shape) == 3: c,w,h = x.shape unshuffled = F.pixel_unshuffle(x, f) if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value) return unshuffled.view(c,f,f,w//f+2_pad,h//f+2_pad).permute(0,1,3,2,4).reshape(c, w+2_f_pad, h+2_f_pad) # batched image tensor else: b,c,w,h = x.shape unshuffled = F.pixel_unshuffle(x, f) if pad != 0: unshuffled = F.pad(unshuffled, (pad, pad, pad, pad), value=pad_value) return unshuffled.view(b,c,f,f,w//f+2_pad,h//f+2_pad).permute(0,1,2,4,3,5).reshape(b,c,w+2_f_pad, h+2_f_pad)
 
do PD
     if pd &gt; 1:         pd_img = pixel_shuffle_down_sampling(img, f=pd, pad=self.pd_pad)  
以上是将图片输入到APBSN之前的PD操作代码，对于(8,3,120,120)的输入在pd=5后输出依然是(8,3,120,120)，而通过您的RSG操作后形状会变成25个(8,3,24,24),我想知道的是即便相对于原版的APBSN参数量是一样的，但是运算次数变为25（每次运算的参数变为1/25），这样也会使训练时间变长对吗(感觉训练时间多了10-20倍)，辛苦您的答复
  
是的，就是这里导致的训练时间变长（并行要比串行快很多）
